sql/src/main/java/org/apache/druid/sql/calcite/rule/FlattenConcatRule.java

```diff
@@ -0,0 +1,125 @@
+import org.apache.druid.math.expr.Function;
+import org.apache.druid.sql.calcite.expression.builtin.ConcatOperatorConversion;
+import org.apache.druid.sql.calcite.expression.builtin.TextcatOperatorConversion;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Flattens calls to CONCAT. Useful because otherwise [a || b || c] would get planned as [CONCAT(CONCAT(a, b), c)].
+ */
+public class FlattenConcatRule extends RelOptRule implements SubstitutionRule
+{
+  public FlattenConcatRule()
+  {
+    super(operand(RelNode.class, any()));
+  }
+
+  @Override
+  public void onMatch(RelOptRuleCall call)
+  {
+    final RelNode oldNode = call.rel(0);
+    final FlattenConcatShuttle shuttle = new FlattenConcatShuttle(oldNode.getCluster().getRexBuilder());
+    final RelNode newNode = oldNode.accept(shuttle);
+
+    //noinspection ObjectEquality
+    if (newNode != oldNode) {
+      call.transformTo(newNode);
+      call.getPlanner().prune(oldNode);
+    }
+  }
+
+  private static class FlattenConcatShuttle extends RexShuttle
+  {
+    private final RexBuilder rexBuilder;
+
+    public FlattenConcatShuttle(RexBuilder rexBuilder)
+    {
+      this.rexBuilder = rexBuilder;
+    }
+
+    @Override
+    public RexNode visitCall(RexCall call)
+    {
+      if (isNonTrivialStringConcat(call)) {
+        final List<RexNode> newOperands = new ArrayList<>();
+        for (final RexNode operand : call.getOperands()) {
+          if (isStringConcat(operand)) {
+            newOperands.addAll(((RexCall) visitCall((RexCall) operand)).getOperands());
+          } else if (RexUtil.isNullLiteral(operand, true) && NullHandling.sqlCompatible()) {
+            return rexBuilder.makeNullLiteral(call.getType());
+          } else {
+            newOperands.add(operand);
+          }
+        }
+
+        if (!newOperands.equals(call.getOperands())) {
+          return rexBuilder.makeCall(ConcatOperatorConversion.SQL_FUNCTION, newOperands);
+        } else {
+          return call;
+        }
+      } else {
+        return super.visitCall(call);
+      }
+    }
+  }
+
+  /**
+   * Whether a rex is a string concatenation operator. All of these end up being converted to
+   * {@link Function.ConcatFunc}.
+   */
+  static boolean isStringConcat(final RexNode rexNode)
+  {
+    if (SqlTypeFamily.STRING.contains(rexNode.getType()) && rexNode instanceof RexCall) {
+      final SqlOperator operator = ((RexCall) rexNode).getOperator();
+      return ConcatOperatorConversion.SQL_FUNCTION.equals(operator)
+             || TextcatOperatorConversion.SQL_FUNCTION.equals(operator)
+             || SqlStdOperatorTable.CONCAT.equals(operator);
+    } else {
+      return false;
+    }
+  }
+
+  /**
+   * Whether a rex is a string concatenation involving at least one an input field.
+   */
+  static boolean isNonTrivialStringConcat(final RexNode rexNode)
+  {
+    return FlattenConcatRule.isStringConcat(rexNode) && !RelOptUtil.InputFinder.bits(rexNode).isEmpty();
```

Comment:
I wonder why restrict to only handle at least 1 input field? 
I believe handling:
```
concat( x , concat( concat ( 'a' , 'b'), 'c') )
```
wouldn't be usefull? or that might be handled elsewhere?

Reply:
If there are zero input fields, like `concat ( 'a' , 'b')`, that would be reduced to `'ab'` by ReduceExpressionRule, which is even better than flattening it in to a parent concat. I added a comment about this, and added a test as well (`testConcat2`).